File: Datasets/TrickyBugs/GenProgs/dpp_generated_progs_cpp/p03299/p03299_num3_parsed.cpp
#include <iostream>
#include <vector>

using namespace std;

const int MOD = 1000000007;

int main() {
    int N;
    cin >> N;
    
    vector<int> h(N);
    for (int i = 0; i < N; i++) {
        cin >> h[i];
    }
    
    vector<vector<int>> dp(N, vector<int>(2));
    dp[0][0] = 1;
    dp[0][1] = 1;
    
    for (int i = 1; i < N; i++) {
        if (h[i] >= h[i-1]) {
            dp[i][0] = (dp[i-1][0] + dp[i-1][1]) % MOD;
        }
        if (h[i] <= h[i-1]) {
            dp[i][1] = (dp[i-1][0] + dp[i-1][1]) % MOD;
        }
    }
    
    int result = (dp[N-1][0] + dp[N-1][1]) % MOD;
    cout << result << endl;
    
    return 0;
}

File: Datasets/TrickyBugs/GenProgs/dpp_generated_progs_python/p03299/p03299_num3_parsed.py
import sys

def count_ways(N, heights):
    MOD = 1000000007

    dp = [[0] * (N+1) for _ in range(N+1)]
    dp[0][0] = 1

    for i in range(N):
        for j in range(N+1):
            if j > heights[i]:
                dp[i+1][j] = (dp[i+1][j] + dp[i][j-1]) % MOD
            if j < heights[i]:
                dp[i+1][j] = (dp[i+1][j] + dp[i][j+1]) % MOD
            dp[i+1][j] = (dp[i+1][j] + dp[i][j]) % MOD

    return dp[N][0]

def main():
    N = int(sys.stdin.readline())
    heights = list(map(int, sys.stdin.readline().split()))

    result = count_ways(N, heights)
    print(result)

if __name__ == '__main__':
    main()